lib/vfscore/main.c

```diff
@@ -1164,39 +1179,93 @@ struct dirent *readdir(DIR *dir)
 
 int readdir_r(DIR *dir, struct dirent *entry, struct dirent **result)
 {
-	int error;
+	/**
+	 * `sys_readdir()` will end up calling `VOP_READDIR`, which expects
+	 * a dirent64 structure instead of the `struct dirent` entry parameter.
+	 */
+	struct dirent64 entry64;
 	struct vfscore_file *fp;
+	int error;
 
 	trace_vfs_readdir(dir->fd, entry);
 	error = fget(dir->fd, &fp);
-	if (!error) {
-		error = sys_readdir(fp, entry);
-		fdrop(fp);
-		if (error) {
-			trace_vfs_readdir_err(error);
-		} else {
-			trace_vfs_readdir_ret();
-		}
+	if (unlikely(error))
+		goto err_out;
+
+	error = sys_readdir(fp, &entry64);
+	fdrop(fp);
+	if (unlikely(error)) {
+		trace_vfs_readdir_err(error);
+		goto err_out;
+	} else {
+		trace_vfs_readdir_ret();
 	}
+
+	entry->d_ino = entry64.d_ino;
+	entry->d_type = entry64.d_type;
+	entry->d_off = entry64.d_off;
+	memcpy(entry->d_name, entry64.d_name, sizeof(entry64.d_name));
+
+err_out:
 	// Our dirent has (like Linux) a d_reclen field, but a constant size.
 	entry->d_reclen = sizeof(*entry);
 
-	if (error) {
+	if (unlikely(error))
 		*result = NULL;
-	} else {
+	else
 		*result = entry;
-	}
+
 	return error == ENOENT ? 0 : error;
 }
+#else
+struct dirent *readdir(DIR *dir) __attribute__((alias("readdir64")));
+int readdir_r(DIR *dir, struct dirent *entry, struct dirent **result)
+	 __attribute__((alias("readdir64_r")));
+#endif /* CONFIG_NONLARGEFILE_SYSCALLS */
 
-// FIXME: in 64bit dirent64 and dirent are identical, so it's safe to alias
 #undef readdir64_r
 int readdir64_r(DIR *dir, struct dirent64 *entry,
 		struct dirent64 **result)
-		__attribute__((alias("readdir_r")));
+{
+	struct vfscore_file *fp;
+	int error;
+
+	trace_vfs_readdir64(dir->fd, entry);
+	error = fget(dir->fd, &fp);
+	if (!error) {
+		error = sys_readdir(fp, entry);
+		fdrop(fp);
+		if (unlikely(error))
+			trace_vfs_readdir_err64(error);
+		else
+			trace_vfs_readdir_ret64();
+	}
+	// Our dirent has (like Linux) a d_reclen field, but a constant size.
+	entry->d_reclen = sizeof(*entry);
+
+	if (unlikely(error))
+		*result = NULL;
+	else
+		*result = entry;
+
+	return error == ENOENT ? 0 : error;
+}
 
 #undef readdir64
-struct dirent *readdir64(DIR *dir) __attribute__((alias("readdir")));
+struct dirent64 *readdir64(DIR *dir)
+{
+	static __thread struct dirent64 entry, *result;
+	int ret;
```

Comment:
Same here. What do you think of this :thinking::
```C
struct dirent64 *readdir64(DIR *dir)
{
	static __thread struct dirent64 entry, *result;

        /* By definition, `readdir64_r` already places NULL into `result` if end of directory reached */
	errno = readdir64_r(dir, &entry, &result);

	return result;
}
```

Reply:
Oh, one more thing, does `result` really need to be in the TLS here, I am not really sure of that? There is already `entry` keeping track of this, is that not enough :thinking:.

Reply:
Fair, I don't think that's needed there. I just copy pasted the implementation we have for [`readdir`](https://github.com/unikraft/unikraft/blob/staging/lib/vfscore/main.c#L1150), so no mind filter involved :laughing: 
I'll remove `result` from the TLS.

Reply:
I know what that's like :rofl:. So `readdir` is flawed too :laughing:? This is going to look very inconsistent then, perhaps you should add an additional commit changing that? So, first commit of the PR the one that changes the existing `readdir` removing `result` from TLS and second commit of the PR the one you have right now :1234:. What do you think?

Reply:
Yes, I'll do that :laughing: